# Explore PUDL FERC Form 2 Data

In [1]:
import sqlite3
import pandas as pd

In [2]:
# Set pandas options
pd.set_option('display.max_columns', None)

## Purpose

The purpose of this notebook will be to explore how to use Pythonb to extract [(FERC)](https://ferc.gov/what-ferc) Form 2 data provided by the Public Utility Data Liberation [(PUDL)](https://catalyst.coop/pudl/) Project.

The goal will be to extract Income Statement data for Montana Gas pipelines from the PUDL Sqlite database and match to the FERC Form 2 exported from the FERC website.

The following natural gas pipelines operate in Montna and file a FERC Form 2:
- Colorado Interstate Gas Company L.L.C.
- Northern Border Pipeline Company
- WBI Energy Transmission, Inc.

## Establish Connection to PUDL Database

The first step will be to connect to the SQLite database downloaded from the PUDL website.

In [3]:
# Connect to the database
conn = sqlite3.connect("ferc2_xbrl.sqlite")

In [4]:
# Create a cursor that lets you execute SQL queries and fetch results
cursor = conn.cursor()

In [5]:
# Get list of table names
cursor.execute("SELECT name FROM sqlite_master WHERE type='table';")

## Explore Tables Contained in the Database

The PUDL SQLite database conatins many tables, each representing various schedules and statements. With the database loaded we can now extract a list of the various tables available.

In [6]:
# Assign list of tables to variable
tables = cursor.fetchall()

In [7]:
# Print list of table names
print("Tables in the database:")
for table in tables:
    print(table[0])

Tables in the database:
corporate_officer_certification_001_duration
identification_001_duration
list_of_schedules_002_duration
general_information_101_duration
control_over_respondent_102_duration
control_over_respondent_102_instant
corporations_controlled_by_respondent_103_duration
corporations_controlled_by_respondent_103_instant
security_holders_and_voting_powers_107_duration
security_holders_and_voting_powers_107_instant
security_holders_and_voting_powers_sequence_107_duration
security_holders_and_voting_powers_sequence_107_instant
important_changes_during_the_quarter_year_108_1_duration
comparative_balance_sheet_assets_and_other_debits_110_instant
comparative_balance_sheet_liabilities_and_other_credits_110_instant
statement_of_income_114_duration
statement_of_accumulated_other_comprehensive_income_and_hedging_activities_117_duration
statement_of_accumulated_other_comprehensive_income_and_hedging_activities_117_instant
retained_earnings_118_duration
retained_earnings_118_instant
r

## Extract Income Statement Data

Below are screenshots of the FERC Form 2 Stateemnt of Income for Colorado Interstate Gas Company:

![](img/ferc-form2-income-statement-01.png)

![](img/ferc-form2-income-statement-02.png)

![](img/ferc-form2-income-statement-03.png)

By looking at the list of tables from earlier, "statement_of_income_114_duration" would appear to be the relevant table name as the FERC Form 2, Statement of Income is on Page 114 of the form. To verify this we can assign this table to a Pandas DataFrame to explore the table columns.

In [8]:
# Assign statement of inocme table to a DataFrame then close the sql connection
df_is = pd.read_sql("SELECT * FROM statement_of_income_114_duration", conn)
conn.close()

In [9]:
df_is.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3011 entries, 0 to 3010
Data columns (total 74 columns):
 #   Column                                                                        Non-Null Count  Dtype  
---  ------                                                                        --------------  -----  
 0   entity_id                                                                     3011 non-null   object 
 1   filing_name                                                                   3011 non-null   object 
 2   publication_time                                                              3011 non-null   object 
 3   start_date                                                                    3011 non-null   object 
 4   end_date                                                                      3011 non-null   object 
 5   utility_type_axis                                                             3011 non-null   object 
 6   revenues_from_merchandising_jobb

In [10]:
# Printout of first 20 entity_id's and filing_name's
df_is.loc[:, ('entity_id', 'filing_name')].head(20)

entity_id                                        filing_name
0    C000251     Trunkline_Gas_Company,_LLC_form2_Q4_1649968474
1    C000251     Trunkline_Gas_Company,_LLC_form2_Q4_1649968474
2    C000251     Trunkline_Gas_Company,_LLC_form2_Q4_1649968474
3    C000251     Trunkline_Gas_Company,_LLC_form2_Q4_1649968474
4    C000251     Trunkline_Gas_Company,_LLC_form2_Q4_1649968474
5    C000251     Trunkline_Gas_Company,_LLC_form2_Q4_1649968474
6    C000251     Trunkline_Gas_Company,_LLC_form2_Q4_1649968474
7    C000251     Trunkline_Gas_Company,_LLC_form2_Q4_1649968474
8    C000251     Trunkline_Gas_Company,_LLC_form2_Q4_1649968474
9    C001157  Cameron_Interstate_Pipeline,_LLC_form2_Q4_1650...
10   C001157  Cameron_Interstate_Pipeline,_LLC_form2_Q4_1650...
11   C001157  Cameron_Interstate_Pipeline,_LLC_form2_Q4_1650...
12   C001157  Cameron_Interstate_Pipeline,_LLC_form2_Q4_1650...
13   C001157  Cameron_Interstate_Pipeline,_LLC_form2_Q4_1650...
14   C001157  Cameron_Interstate_Pipeline,_LLC_form2_Q4_1650...
15   C001157  Cameron_Interstate_Pipeline,_LLC_form2_Q4_1650...
16   C000097   Cimarron_River_Pipeline,_LLC_form2_Q4_1652892936
17   C000097   Cimarron_River_Pipeline,_LLC_form2_Q4_1652892936
18   C000097   Cimarron_River_Pipeline,_LLC_form2_Q4_1652892936
19   C000097   Cimarron_River_Pipeline,_LLC_form2_Q4_1652892936

In [11]:
# Get count of unique entity_id's
unique_entity_count = df_is['entity_id'].nunique()
print(f'Count of unique entities: {unique_entity_count}')

Count of unique entities: 106


The initial DataFrame contains thousands of rows of data for 106 unique filers. For this example, we will start with pulling information for Colorado Interstate Gas Company (CIGC). To be able to do this the DataFrame will need to be filtered. The **entity_id** column contains a unique identifier for each filer and **filing_name** contains the name of the filer. The **filing_name** column can be used to determine which **entity_id** corresponds to CIGC.

In [12]:
# Use str.contains to find CIGC's entity id
df_is[df_is.loc[:, 'filing_name'].str.contains('Colorado_Interstate')]

entity_id                                        filing_name  \
54     C000981  Colorado_Interstate_Gas_Company,_L.L.C._form2_...   
55     C000981  Colorado_Interstate_Gas_Company,_L.L.C._form2_...   
56     C000981  Colorado_Interstate_Gas_Company,_L.L.C._form2_...   
57     C000981  Colorado_Interstate_Gas_Company,_L.L.C._form2_...   
58     C000981  Colorado_Interstate_Gas_Company,_L.L.C._form2_...   
59     C000981  Colorado_Interstate_Gas_Company,_L.L.C._form2_...   
60     C000981  Colorado_Interstate_Gas_Company,_L.L.C._form2_...   
1253   C000981  Colorado_Interstate_Gas_Company,_L.L.C._form2_...   
1254   C000981  Colorado_Interstate_Gas_Company,_L.L.C._form2_...   
1255   C000981  Colorado_Interstate_Gas_Company,_L.L.C._form2_...   
1256   C000981  Colorado_Interstate_Gas_Company,_L.L.C._form2_...   
1257   C000981  Colorado_Interstate_Gas_Company,_L.L.C._form2_...   
1258   C000981  Colorado_Interstate_Gas_Company,_L.L.C._form2_...   
1259   C000981  Colorado_Interstate_Gas_Company,_L.L.C._form2_...   
2288   C000981  Colorado_Interstate_Gas_Company,_L.L.C._form2_...   
2289   C000981  Colorado_Interstate_Gas_Company,_L.L.C._form2_...   
2290   C000981  Colorado_Interstate_Gas_Company,_L.L.C._form2_...   
2291   C000981  Colorado_Interstate_Gas_Company,_L.L.C._form2_...   
2292   C000981  Colorado_Interstate_Gas_Company,_L.L.C._form2_...   
2293   C000981  Colorado_Interstate_Gas_Company,_L.L.C._form2_...   
2294   C000981  Colorado_Interstate_Gas_Company,_L.L.C._form2_...   
2814   C000981  Colorado_Interstate_Gas_Company,_L.L.C._form2_...   
2815   C000981  Colorado_Interstate_Gas_Company,_L.L.C._form2_...   
2816   C000981  Colorado_Interstate_Gas_Company,_L.L.C._form2_...   
2817   C000981  Colorado_Interstate_Gas_Company,_L.L.C._form2_...   
2818   C000981  Colorado_Interstate_Gas_Company,_L.L.C._form2_...   
2819   C000981  Colorado_Interstate_Gas_Company,_L.L.C._form2_...   
2820   C000981  Colorado_Interstate_Gas_Company,_L.L.C._form2_...   

                publication_time  start_date    end_date  \
54    2022-04-01 02:43:53.000000  2021-01-01  2021-12-31   
55    2022-04-01 02:43:53.000000  2020-01-01  2020-12-31   
56    2022-04-01 02:43:53.000000  2021-10-01  2021-10-31   
57    2022-04-01 02:43:53.000000  2021-11-01  2021-11-30   
58    2022-04-01 02:43:53.000000  2021-12-01  2021-12-31   
59    2022-04-01 02:43:53.000000  2021-01-01  2021-12-31   
60    2022-04-01 02:43:53.000000  2020-01-01  2020-12-31   
1253  2023-03-30 14:09:00.000000  2022-01-01  2022-12-31   
1254  2023-03-30 14:09:00.000000  2022-01-01  2022-12-31   
1255  2023-03-30 14:09:00.000000  2021-01-01  2021-12-31   
1256  2023-03-30 14:09:00.000000  2021-01-01  2021-12-31   
1257  2023-03-30 14:09:00.000000  2022-10-01  2022-10-31   
1258  2023-03-30 14:09:00.000000  2022-11-01  2022-11-30   
1259  2023-03-30 14:09:00.000000  2022-12-01  2022-12-31   
2288  2024-03-29 12:24:32.000000  2023-01-01  2023-12-31   
2289  2024-03-29 12:24:32.000000  2022-01-01  2022-12-31   
2290  2024-03-29 12:24:32.000000  2023-01-01  2023-12-31   
2291  2024-03-29 12:24:32.000000  2022-01-01  2022-12-31   
2292  2024-03-29 12:24:32.000000  2023-10-01  2023-10-31   
2293  2024-03-29 12:24:32.000000  2023-11-01  2023-11-30   
2294  2024-03-29 12:24:32.000000  2023-12-01  2023-12-31   
2814  2025-03-27 22:02:24.000000  2024-01-01  2024-12-31   
2815  2025-03-27 22:02:24.000000  2024-01-01  2024-12-31   
2816  2025-03-27 22:02:24.000000  2023-01-01  2023-12-31   
2817  2025-03-27 22:02:24.000000  2023-01-01  2023-12-31   
2818  2025-03-27 22:02:24.000000  2024-10-01  2024-10-31   
2819  2025-03-27 22:02:24.000000  2024-11-01  2024-11-30   
2820  2025-03-27 22:02:24.000000  2024-12-01  2024-12-31   

          utility_type_axis  \
54                    total   
55                    total   
56                    total   
57                    total   
58                    total   
59    ferc:GasUtilityMember   
60    ferc:GasUtilityMember   
1253          

Colorado Interstate Gas Company's entity id is **C000981**. We can now use this ID to filter the DataFrame for CIGC.

In [13]:
# Use CIGC's entity id to filter the DataFrame
id_cigc = 'C000981'
df_is_cigc = df_is[df_is['entity_id'] == id_cigc]

In [14]:
df_is_cigc

entity_id                                        filing_name  \
54     C000981  Colorado_Interstate_Gas_Company,_L.L.C._form2_...   
55     C000981  Colorado_Interstate_Gas_Company,_L.L.C._form2_...   
56     C000981  Colorado_Interstate_Gas_Company,_L.L.C._form2_...   
57     C000981  Colorado_Interstate_Gas_Company,_L.L.C._form2_...   
58     C000981  Colorado_Interstate_Gas_Company,_L.L.C._form2_...   
59     C000981  Colorado_Interstate_Gas_Company,_L.L.C._form2_...   
60     C000981  Colorado_Interstate_Gas_Company,_L.L.C._form2_...   
1253   C000981  Colorado_Interstate_Gas_Company,_L.L.C._form2_...   
1254   C000981  Colorado_Interstate_Gas_Company,_L.L.C._form2_...   
1255   C000981  Colorado_Interstate_Gas_Company,_L.L.C._form2_...   
1256   C000981  Colorado_Interstate_Gas_Company,_L.L.C._form2_...   
1257   C000981  Colorado_Interstate_Gas_Company,_L.L.C._form2_...   
1258   C000981  Colorado_Interstate_Gas_Company,_L.L.C._form2_...   
1259   C000981  Colorado_Interstate_Gas_Company,_L.L.C._form2_...   
2288   C000981  Colorado_Interstate_Gas_Company,_L.L.C._form2_...   
2289   C000981  Colorado_Interstate_Gas_Company,_L.L.C._form2_...   
2290   C000981  Colorado_Interstate_Gas_Company,_L.L.C._form2_...   
2291   C000981  Colorado_Interstate_Gas_Company,_L.L.C._form2_...   
2292   C000981  Colorado_Interstate_Gas_Company,_L.L.C._form2_...   
2293   C000981  Colorado_Interstate_Gas_Company,_L.L.C._form2_...   
2294   C000981  Colorado_Interstate_Gas_Company,_L.L.C._form2_...   
2814   C000981  Colorado_Interstate_Gas_Company,_L.L.C._form2_...   
2815   C000981  Colorado_Interstate_Gas_Company,_L.L.C._form2_...   
2816   C000981  Colorado_Interstate_Gas_Company,_L.L.C._form2_...   
2817   C000981  Colorado_Interstate_Gas_Company,_L.L.C._form2_...   
2818   C000981  Colorado_Interstate_Gas_Company,_L.L.C._form2_...   
2819   C000981  Colorado_Interstate_Gas_Company,_L.L.C._form2_...   
2820   C000981  Colorado_Interstate_Gas_Company,_L.L.C._form2_...   

                publication_time  start_date    end_date  \
54    2022-04-01 02:43:53.000000  2021-01-01  2021-12-31   
55    2022-04-01 02:43:53.000000  2020-01-01  2020-12-31   
56    2022-04-01 02:43:53.000000  2021-10-01  2021-10-31   
57    2022-04-01 02:43:53.000000  2021-11-01  2021-11-30   
58    2022-04-01 02:43:53.000000  2021-12-01  2021-12-31   
59    2022-04-01 02:43:53.000000  2021-01-01  2021-12-31   
60    2022-04-01 02:43:53.000000  2020-01-01  2020-12-31   
1253  2023-03-30 14:09:00.000000  2022-01-01  2022-12-31   
1254  2023-03-30 14:09:00.000000  2022-01-01  2022-12-31   
1255  2023-03-30 14:09:00.000000  2021-01-01  2021-12-31   
1256  2023-03-30 14:09:00.000000  2021-01-01  2021-12-31   
1257  2023-03-30 14:09:00.000000  2022-10-01  2022-10-31   
1258  2023-03-30 14:09:00.000000  2022-11-01  2022-11-30   
1259  2023-03-30 14:09:00.000000  2022-12-01  2022-12-31   
2288  2024-03-29 12:24:32.000000  2023-01-01  2023-12-31   
2289  2024-03-29 12:24:32.000000  2022-01-01  2022-12-31   
2290  2024-03-29 12:24:32.000000  2023-01-01  2023-12-31   
2291  2024-03-29 12:24:32.000000  2022-01-01  2022-12-31   
2292  2024-03-29 12:24:32.000000  2023-10-01  2023-10-31   
2293  2024-03-29 12:24:32.000000  2023-11-01  2023-11-30   
2294  2024-03-29 12:24:32.000000  2023-12-01  2023-12-31   
2814  2025-03-27 22:02:24.000000  2024-01-01  2024-12-31   
2815  2025-03-27 22:02:24.000000  2024-01-01  2024-12-31   
2816  2025-03-27 22:02:24.000000  2023-01-01  2023-12-31   
2817  2025-03-27 22:02:24.000000  2023-01-01  2023-12-31   
2818  2025-03-27 22:02:24.000000  2024-10-01  2024-10-31   
2819  2025-03-27 22:02:24.000000  2024-11-01  2024-11-30   
2820  2025-03-27 22:02:24.000000  2024-12-01  2024-12-31   

          utility_type_axis  \
54                    total   
55                    total   
56                    total   
57                    total   
58                    total   
59    ferc:GasUtilityMember   
60    ferc:GasUtilityMember   
1253          

From reviewing the above there will be overlap of data due to the fact that for each FERC form there is current and prior year data reported. This results in prior year data being duplicated. For this section of the analyis our goal is to match the current year FERC Form 2, so we will further filter the data to the most recent form. To do this we can filter using the **publication_time** column and only keep rows published in 2025.

In [15]:
# Filter the DataFrame by publication_time
df_is_cigc_25 = df_is_cigc[df_is_cigc.loc[:, 'publication_time'].str.contains('2025-')]

In [16]:
df_is_cigc_25

entity_id                                        filing_name  \
2814   C000981  Colorado_Interstate_Gas_Company,_L.L.C._form2_...   
2815   C000981  Colorado_Interstate_Gas_Company,_L.L.C._form2_...   
2816   C000981  Colorado_Interstate_Gas_Company,_L.L.C._form2_...   
2817   C000981  Colorado_Interstate_Gas_Company,_L.L.C._form2_...   
2818   C000981  Colorado_Interstate_Gas_Company,_L.L.C._form2_...   
2819   C000981  Colorado_Interstate_Gas_Company,_L.L.C._form2_...   
2820   C000981  Colorado_Interstate_Gas_Company,_L.L.C._form2_...   

                publication_time  start_date    end_date  \
2814  2025-03-27 22:02:24.000000  2024-01-01  2024-12-31   
2815  2025-03-27 22:02:24.000000  2024-01-01  2024-12-31   
2816  2025-03-27 22:02:24.000000  2023-01-01  2023-12-31   
2817  2025-03-27 22:02:24.000000  2023-01-01  2023-12-31   
2818  2025-03-27 22:02:24.000000  2024-10-01  2024-10-31   
2819  2025-03-27 22:02:24.000000  2024-11-01  2024-11-30   
2820  2025-03-27 22:02:24.000000  2024-12-01  2024-12-31   

          utility_type_axis  \
2814                  total   
2815  ferc:GasUtilityMember   
2816  ferc:GasUtilityMember   
2817                  total   
2818                  total   
2819                  total   
2820                  total   

      revenues_from_merchandising_jobbing_and_contract_work  \
2814                                                NaN       
2815                                                NaN       
2816                                                NaN       
2817                                                NaN       
2818                                                NaN       
2819                                                NaN       
2820                                                NaN       

      losses_from_disposition_of_allowances  donations  maintenance_expense  \
2814                                    NaN        NaN           32979655.0   
2815                                    NaN        NaN           32979655.0   
2816                                    NaN        NaN           28854725.0   
2817                                    NaN        NaN           28854725.0   
2818                                    NaN        NaN                  NaN   
2819                                    NaN        NaN                  NaN   
2820                                    NaN        NaN                  NaN   

      depreciation_expense_for_asset_retirement_costs  \
2814                                              NaN   
2815                                              NaN   
2816                                              NaN   
2817                                              NaN   
2818                                              NaN   
2819                                              NaN   
2820                                              NaN   

      net_other_income_and_deductions  \
2814                       36381297.0   
2815                              NaN   
2816                              NaN   
2817                       36502362.0   
2818                              NaN   
2819                              NaN   
2820                              NaN   

      amortization_and_depletion_of_utility_plant  \
2814                                     165213.0   
2815                                     165213.0   
2816                                     224340.0   
2817                                     224340.0   
2818                                          NaN   
2819                                          NaN   
2820                                          NaN   

      income_taxes_utility_operating_income_other  other_income  \
2814                                    1747527.0    47620477.0   
2815                                    1747527.0           NaN   
2816                                    1871571.0           NaN   
2817                                    1871571.0    47780754.0   
2818                                  

From a quick review this does appear to be the correct information. However, the column orders do not match to the FERC Form 2. To be able to complete a full comparison of the data the columns are going to need to be rearanged in an order that matches to the Form 2. It's also worth noting that there are rows for current and prior year totals as well as current and prior year rows for Elec. Utility, Gas Utility, and Other Utility.

In [17]:
# Get a list of sorted column names
column_names = df_is_cigc_25.columns
column_names = column_names.sort_values()

In [18]:
# Print sorted list
for column in column_names:
    print(column)

accretion_expense
allowance_for_borrowed_funds_used_during_construction_credit
allowance_for_other_funds_used_during_construction
amortization_and_depletion_of_utility_plant
amortization_of_conversion_expenses
amortization_of_debt_discount_and_expense
amortization_of_gain_on_reacquired_debt_credit
amortization_of_gas_plant_acquisition_adjustments
amortization_of_loss_on_reacquired_debt
amortization_of_premium_on_debt_credit
amortization_of_property_losses_unrecovered_plant_and_regulatory_study_costs
costs_and_expenses_of_merchandising_jobbing_and_contract_work
depreciation_expense
depreciation_expense_for_asset_retirement_costs
donations
end_date
entity_id
equity_in_earnings_of_subsidiary_companies
expenditures_for_certain_civic_political_and_related_activities
expenses_of_nonutility_operations
extraordinary_deductions
extraordinary_income
extraordinary_items_after_taxes
filing_name
gain_on_disposition_of_property
gains_from_disposition_of_allowances
gains_from_disposition_of_plant
inc

This sorted list was then used to compare to the XBRL form 2 available from the FERC website for Coloraod Interstate Gas Company. The columns match to a specific xbrl tag that correlates with an item in the statements. For example the below screenshot shows the xbrl tag for the first item in the statement, operating revenues:

![](img/ferc-form2-xbrl-tag.png)

The XBRL tag "OperatingReveneus" correlates with column "operating_revenues" in the PUDL database. \
\
By looking up the tag and matching to the list of column names, a list was put together that allowed the database to be sorted in the same order as the Statement of Income.

In [19]:
# Produce a printout of the DataFrame in an order that matches to the Income Statement
df_is_cigc_25.loc[:, [
    'entity_id',
    'filing_name',
    'publication_time',
    'utility_type_axis',
    'start_date',
    'end_date',
    'operating_revenues',
    'operation_expense',
    'maintenance_expense',
    'depreciation_expense',
    'depreciation_expense_for_asset_retirement_costs',
    'amortization_and_depletion_of_utility_plant',
    'amortization_of_gas_plant_acquisition_adjustments',
    'amortization_of_property_losses_unrecovered_plant_and_regulatory_study_costs',
    'amortization_of_conversion_expenses',
    'regulatory_debits',
    'regulatory_credits',
    'taxes_other_than_income_taxes_utility_operating_income',
    'income_taxes_utility_operating_income',
    'income_taxes_utility_operating_income_other',
    'provisions_for_deferred_income_taxes_utility_operating_income',
    'provision_for_deferred_income_taxes_credit_utility_operating_income',
    'investment_tax_credit_adjustments',
    'gains_from_disposition_of_plant',
    'losses_from_disposition_of_utility_plant',
    'gains_from_disposition_of_allowances',
    'losses_from_disposition_of_allowances',
    'accretion_expense',
    'utility_operating_expenses',
    'net_utility_operating_income',
    'revenues_from_merchandising_jobbing_and_contract_work',
    'costs_and_expenses_of_merchandising_jobbing_and_contract_work',
    'revenues_from_nonutility_operations',
    'expenses_of_nonutility_operations',
    'nonoperating_rental_income',
    'equity_in_earnings_of_subsidiary_companies',
    'interest_and_dividend_income',
    'allowance_for_other_funds_used_during_construction',
    'miscellaneous_nonoperating_income',
    'gain_on_disposition_of_property',
    'other_income',
    'loss_on_disposition_of_property',
    'miscellaneous_amortization',
    'donations',
    'life_insurance',
    'penalties',
    'expenditures_for_certain_civic_political_and_related_activities',
    'other_deductions',
    'other_income_deductions',
    'taxes_other_than_income_taxes_other_income_and_deductions',
    'income_taxes_federal',
    'income_taxes_other',
    'provision_for_deferred_income_taxes_other_income_and_deductions',
    'provision_for_deferred_income_taxes_credit_other_income_and_deductions',
    'investment_tax_credit_adjustments_nonutility_operations',
    'investment_tax_credits',
    'taxes_on_other_income_and_deductions',
    'net_other_income_and_deductions',
    'interest_on_long_term_debt',
    'amortization_of_debt_discount_and_expense',
    'amortization_of_loss_on_reacquired_debt',
    'amortization_of_premium_on_debt_credit',
    'amortization_of_gain_on_reacquired_debt_credit',
    'interest_on_debt_to_associated_companies',
    'other_interest_expense',
    'allowance_for_borrowed_funds_used_during_construction_credit',
    'net_interest_charges',
    'income_before_extraordinary_items',
    'extraordinary_income',
    'extraordinary_deductions',
    'net_extraordinary_items',
    'income_taxes_extraordinary_items',
    'extraordinary_items_after_taxes',
    'net_income_loss'
    ]]

entity_id                                        filing_name  \
2814   C000981  Colorado_Interstate_Gas_Company,_L.L.C._form2_...   
2815   C000981  Colorado_Interstate_Gas_Company,_L.L.C._form2_...   
2816   C000981  Colorado_Interstate_Gas_Company,_L.L.C._form2_...   
2817   C000981  Colorado_Interstate_Gas_Company,_L.L.C._form2_...   
2818   C000981  Colorado_Interstate_Gas_Company,_L.L.C._form2_...   
2819   C000981  Colorado_Interstate_Gas_Company,_L.L.C._form2_...   
2820   C000981  Colorado_Interstate_Gas_Company,_L.L.C._form2_...   

                publication_time      utility_type_axis  start_date  \
2814  2025-03-27 22:02:24.000000                  total  2024-01-01   
2815  2025-03-27 22:02:24.000000  ferc:GasUtilityMember  2024-01-01   
2816  2025-03-27 22:02:24.000000  ferc:GasUtilityMember  2023-01-01   
2817  2025-03-27 22:02:24.000000                  total  2023-01-01   
2818  2025-03-27 22:02:24.000000                  total  2024-10-01   
2819  2025-03-27 22:02:24.000000                  total  2024-11-01   
2820  2025-03-27 22:02:24.000000                  total  2024-12-01   

        end_date  operating_revenues  operation_expense  maintenance_expense  \
2814  2024-12-31         332110238.0        143456709.0           32979655.0   
2815  2024-12-31         332110238.0        143456709.0           32979655.0   
2816  2023-12-31         332489669.0        143854809.0           28854725.0   
2817  2023-12-31         332489669.0        143854809.0           28854725.0   
2818  2024-10-31          28201699.0                NaN                  NaN   
2819  2024-11-30          29790470.0                NaN                  NaN   
2820  2024-12-31          30030732.0                NaN                  NaN   

      depreciation_expense  depreciation_expense_for_asset_retirement_costs  \
2814            32360350.0                                              NaN   
2815            32360350.0                                              NaN   
2816            31609006.0                                              NaN   
2817            31609006.0                                              NaN   
2818                   NaN                                              NaN   
2819                   NaN                                              NaN   
2820                   NaN                                              NaN   

      amortization_and_depletion_of_utility_plant  \
2814                                     165213.0   
2815                                     165213.0   
2816                                     224340.0   
2817                                     224340.0   
2818                                          NaN   
2819                                          NaN   
2820                                          NaN   

      amortization_of_gas_plant_acquisition_adjustments  \
2814                                                NaN   
2815                                                NaN   
2816                                                NaN   
2817                                                NaN   
2818                                                NaN   
2819                                                NaN   
2820                                                NaN   

      amortization_of_property_losses_unrecovered_plant_and_regulatory_study_costs  \
2814                                                NaN                              
2815                                                NaN                              
2816                                                NaN                              
2817                                                NaN                              
2818                                                NaN                              
2819                                                NaN                              
2820                                                NaN                              

      amortization_of_

Comparing the sorted DataFrame to the FERC Form 2 for Colorado Interstate Gas does materially match.

## Income Statement Extract Function

With the correct SQL table and the order of columns identified a function can be developed to extract the Statement of Income for each entity id.

In [20]:
# Define the Income Statement function
def get_ferc2_income_statement(db_file, subject_id):

    # Connect to the sql database using sqlite3
    conn = sqlite3.connect(db_file)

    # Query the database file and assign to pandas dataframe
    sql_table = 'statement_of_income_114_duration'
    sql_query = f'SELECT * FROM {sql_table}'
    df = pd.read_sql(sql_query + " WHERE entity_id = ?", conn, params=(subject_id,))
    
    # Close the database connection
    conn.close()

    # Rearange columns to match income statement format
    income_statement_items = (
        'entity_id',
        'filing_name',
        'publication_time',
        'utility_type_axis',
        'start_date',
        'end_date',
        'operating_revenues',
        'operation_expense',
        'maintenance_expense',
        'depreciation_expense',
        'depreciation_expense_for_asset_retirement_costs',
        'amortization_and_depletion_of_utility_plant',
        'amortization_of_gas_plant_acquisition_adjustments',
        'amortization_of_property_losses_unrecovered_plant_and_regulatory_study_costs',
        'amortization_of_conversion_expenses',
        'regulatory_debits',
        'regulatory_credits',
        'taxes_other_than_income_taxes_utility_operating_income',
        'income_taxes_utility_operating_income',
        'income_taxes_utility_operating_income_other',
        'provisions_for_deferred_income_taxes_utility_operating_income',
        'provision_for_deferred_income_taxes_credit_utility_operating_income',
        'investment_tax_credit_adjustments',
        'gains_from_disposition_of_plant',
        'losses_from_disposition_of_utility_plant',
        'gains_from_disposition_of_allowances',
        'losses_from_disposition_of_allowances',
        'accretion_expense',
        'utility_operating_expenses',
        'net_utility_operating_income',
        'revenues_from_merchandising_jobbing_and_contract_work',
        'costs_and_expenses_of_merchandising_jobbing_and_contract_work',
        'revenues_from_nonutility_operations',
        'expenses_of_nonutility_operations',
        'nonoperating_rental_income',
        'equity_in_earnings_of_subsidiary_companies',
        'interest_and_dividend_income',
        'allowance_for_other_funds_used_during_construction',
        'miscellaneous_nonoperating_income',
        'gain_on_disposition_of_property',
        'other_income',
        'loss_on_disposition_of_property',
        'miscellaneous_amortization',
        'donations',
        'life_insurance',
        'penalties',
        'expenditures_for_certain_civic_political_and_related_activities',
        'other_deductions',
        'other_income_deductions',
        'taxes_other_than_income_taxes_other_income_and_deductions',
        'income_taxes_federal',
        'income_taxes_other',
        'provision_for_deferred_income_taxes_other_income_and_deductions',
        'provision_for_deferred_income_taxes_credit_other_income_and_deductions',
        'investment_tax_credit_adjustments_nonutility_operations',
        'investment_tax_credits',
        'taxes_on_other_income_and_deductions',
        'net_other_income_and_deductions',
        'interest_on_long_term_debt',
        'amortization_of_debt_discount_and_expense',
        'amortization_of_loss_on_reacquired_debt',
        'amortization_of_premium_on_debt_credit',
        'amortization_of_gain_on_reacquired_debt_credit',
        'interest_on_debt_to_associated_companies',
        'other_interest_expense',
        'allowance_for_borrowed_funds_used_during_construction_credit',
        'net_interest_charges',
        'income_before_extraordinary_items',
        'extraordinary_income',
        'extraordinary_deductions',
        'net_extraordinary_items',
        'income_taxes_extraordinary_items',
        'extraordinary_items_after_taxes',
        'net_income_loss'
    )
    
    df = df.loc[:, income_statement_items]
    
    return df

We now can test the function using Colorado Interstate Gas entity id "C000981".

In [21]:
# Set function parameters
db_file = 'ferc2_xbrl.sqlite'
subject_id = 'C000981'

In [22]:
# Query the PUDL SQL Database using the function
df_cigc = get_ferc2_income_statement(db_file, subject_id)

In [23]:
df_cigc

entity_id                                        filing_name  \
0    C000981  Colorado_Interstate_Gas_Company,_L.L.C._form2_...   
1    C000981  Colorado_Interstate_Gas_Company,_L.L.C._form2_...   
2    C000981  Colorado_Interstate_Gas_Company,_L.L.C._form2_...   
3    C000981  Colorado_Interstate_Gas_Company,_L.L.C._form2_...   
4    C000981  Colorado_Interstate_Gas_Company,_L.L.C._form2_...   
5    C000981  Colorado_Interstate_Gas_Company,_L.L.C._form2_...   
6    C000981  Colorado_Interstate_Gas_Company,_L.L.C._form2_...   
7    C000981  Colorado_Interstate_Gas_Company,_L.L.C._form2_...   
8    C000981  Colorado_Interstate_Gas_Company,_L.L.C._form2_...   
9    C000981  Colorado_Interstate_Gas_Company,_L.L.C._form2_...   
10   C000981  Colorado_Interstate_Gas_Company,_L.L.C._form2_...   
11   C000981  Colorado_Interstate_Gas_Company,_L.L.C._form2_...   
12   C000981  Colorado_Interstate_Gas_Company,_L.L.C._form2_...   
13   C000981  Colorado_Interstate_Gas_Company,_L.L.C._form2_...   
14   C000981  Colorado_Interstate_Gas_Company,_L.L.C._form2_...   
15   C000981  Colorado_Interstate_Gas_Company,_L.L.C._form2_...   
16   C000981  Colorado_Interstate_Gas_Company,_L.L.C._form2_...   
17   C000981  Colorado_Interstate_Gas_Company,_L.L.C._form2_...   
18   C000981  Colorado_Interstate_Gas_Company,_L.L.C._form2_...   
19   C000981  Colorado_Interstate_Gas_Company,_L.L.C._form2_...   
20   C000981  Colorado_Interstate_Gas_Company,_L.L.C._form2_...   
21   C000981  Colorado_Interstate_Gas_Company,_L.L.C._form2_...   
22   C000981  Colorado_Interstate_Gas_Company,_L.L.C._form2_...   
23   C000981  Colorado_Interstate_Gas_Company,_L.L.C._form2_...   
24   C000981  Colorado_Interstate_Gas_Company,_L.L.C._form2_...   
25   C000981  Colorado_Interstate_Gas_Company,_L.L.C._form2_...   
26   C000981  Colorado_Interstate_Gas_Company,_L.L.C._form2_...   
27   C000981  Colorado_Interstate_Gas_Company,_L.L.C._form2_...   

              publication_time      utility_type_axis  start_date    end_date  \
0   2022-04-01 02:43:53.000000                  total  2021-01-01  2021-12-31   
1   2022-04-01 02:43:53.000000                  total  2020-01-01  2020-12-31   
2   2022-04-01 02:43:53.000000                  total  2021-10-01  2021-10-31   
3   2022-04-01 02:43:53.000000                  total  2021-11-01  2021-11-30   
4   2022-04-01 02:43:53.000000                  total  2021-12-01  2021-12-31   
5   2022-04-01 02:43:53.000000  ferc:GasUtilityMember  2021-01-01  2021-12-31   
6   2022-04-01 02:43:53.000000  ferc:GasUtilityMember  2020-01-01  2020-12-31   
7   2023-03-30 14:09:00.000000                  total  2022-01-01  2022-12-31   
8   2023-03-30 14:09:00.000000  ferc:GasUtilityMember  2022-01-01  2022-12-31   
9   2023-03-30 14:09:00.000000  ferc:GasUtilityMember  2021-01-01  2021-12-31   
10  2023-03-30 14:09:00.000000                  total  2021-01-01  2021-12-31   
11  2023-03-30 14:09:00.000000                  total  2022-10-01  2022-10-31   
12  2023-03-30 14:09:00.000000                  total  2022-11-01  2022-11-30   
13  2023-03-30 14:09:00.000000                  total  2022-12-01  2022-12-31   
14  2024-03-29 12:24:32.000000                  total  2023-01-01  2023-12-31   
15  2024-03-29 12:24:32.000000                  total  2022-01-01  2022-12-31   
16  2024-03-29 12:24:32.000000  ferc:GasUtilityMember  2023-01-01  2023-12-31   
17  2024-03-29 12:24:32.000000  ferc:GasUtilityMember  2022-01-01  2022-12-31   
18  2024-03-29 12:24:32.000000                  total  2023-10-01  2023-10-31   
19  2024-03-29 12:24:32.000000                  total  2023-11-01  2023-11-30   
20  2024-03-29 12:24:32.000000                  total  2023-12-01  2023-12-31   
21  2025-03-27 22:02:24.000000                  total  2024-01-01  2024-12-31   
22  2025-03-27 22:02:24.000000  ferc:GasUtilityMember  2024-01-01  2024-12-31   
23  2025-03-27 22:02:24.000000  ferc:GasUtilityMember  2023-01-01  2023-12-31   
24  2025-03-27 22:02:24.000000    

The functions appears to work as the resulting DataFrame matches to the FERC Form 2 Statement of Income for CIGC. It now can be tested with additional entities.

## Extract Income Statement for Other Entities

This function can now be tested with the remaining Montana gas pipelines identified at the beginning. \
The sql database was reviewed to determine the entity_id's that match to the remaining Montana gas pipelines:
- Northern Border Pipeline Company = Entity ID: C000626 
- WBI Energy Transmission Inc = Entity ID: C000111

In [24]:
# Assign additional entity_id's to variables
id_northern_border_pipeline = 'C000626'
id_wbi_energy_transmission = 'C000111'

In [25]:
# Extract additional income statements using entitiy_id's
df_northern_border_pipeline = get_ferc2_income_statement(db_file, id_northern_border_pipeline)
df_wbi_energy_transmission = get_ferc2_income_statement(db_file, id_wbi_energy_transmission)

In [26]:
df_northern_border_pipeline

entity_id                                        filing_name  \
0    C000626  Northern_Border_Pipeline_Company_form2_Q4_1649...   
1    C000626  Northern_Border_Pipeline_Company_form2_Q4_1649...   
2    C000626  Northern_Border_Pipeline_Company_form2_Q4_1649...   
3    C000626  Northern_Border_Pipeline_Company_form2_Q4_1649...   
4    C000626  Northern_Border_Pipeline_Company_form2_Q4_1649...   
5    C000626  Northern_Border_Pipeline_Company_form2_Q4_1649...   
6    C000626  Northern_Border_Pipeline_Company_form2_Q4_1649...   
7    C000626  Northern_Border_Pipeline_Company_form2_Q4_1681...   
8    C000626  Northern_Border_Pipeline_Company_form2_Q4_1681...   
9    C000626  Northern_Border_Pipeline_Company_form2_Q4_1681...   
10   C000626  Northern_Border_Pipeline_Company_form2_Q4_1681...   
11   C000626  Northern_Border_Pipeline_Company_form2_Q4_1681...   
12   C000626  Northern_Border_Pipeline_Company_form2_Q4_1681...   
13   C000626  Northern_Border_Pipeline_Company_form2_Q4_1681...   
14   C000626  Northern_Border_Pipeline_Company_form2_Q4_1712...   
15   C000626  Northern_Border_Pipeline_Company_form2_Q4_1712...   
16   C000626  Northern_Border_Pipeline_Company_form2_Q4_1712...   
17   C000626  Northern_Border_Pipeline_Company_form2_Q4_1712...   
18   C000626  Northern_Border_Pipeline_Company_form2_Q4_1712...   
19   C000626  Northern_Border_Pipeline_Company_form2_Q4_1712...   
20   C000626  Northern_Border_Pipeline_Company_form2_Q4_1712...   
21   C000626  Northern_Border_Pipeline_Company_form2_Q4_1744...   
22   C000626  Northern_Border_Pipeline_Company_form2_Q4_1744...   
23   C000626  Northern_Border_Pipeline_Company_form2_Q4_1744...   
24   C000626  Northern_Border_Pipeline_Company_form2_Q4_1744...   
25   C000626  Northern_Border_Pipeline_Company_form2_Q4_1744...   
26   C000626  Northern_Border_Pipeline_Company_form2_Q4_1744...   
27   C000626  Northern_Border_Pipeline_Company_form2_Q4_1744...   

              publication_time      utility_type_axis  start_date    end_date  \
0   2022-04-06 21:06:10.000000  ferc:GasUtilityMember  2020-01-01  2020-12-31   
1   2022-04-06 21:06:10.000000                  total  2021-11-01  2021-11-30   
2   2022-04-06 21:06:10.000000                  total  2021-01-01  2021-12-31   
3   2022-04-06 21:06:10.000000                  total  2021-10-01  2021-10-31   
4   2022-04-06 21:06:10.000000  ferc:GasUtilityMember  2021-01-01  2021-12-31   
5   2022-04-06 21:06:10.000000                  total  2021-12-01  2021-12-31   
6   2022-04-06 21:06:10.000000                  total  2020-01-01  2020-12-31   
7   2023-04-14 14:42:54.000000                  total  2022-11-01  2022-11-30   
8   2023-04-14 14:42:54.000000                  total  2022-10-01  2022-10-31   
9   2023-04-14 14:42:54.000000                  total  2021-01-01  2021-12-31   
10  2023-04-14 14:42:54.000000  ferc:GasUtilityMember  2021-01-01  2021-12-31   
11  2023-04-14 14:42:54.000000  ferc:GasUtilityMember  2022-01-01  2022-12-31   
12  2023-04-14 14:42:54.000000                  total  2022-12-01  2022-12-31   
13  2023-04-14 14:42:54.000000                  total  2022-01-01  2022-12-31   
14  2024-04-08 22:41:48.000000                  total  2023-01-01  2023-12-31   
15  2024-04-08 22:41:48.000000                  total  2022-01-01  2022-12-31   
16  2024-04-08 22:41:48.000000  ferc:GasUtilityMember  2023-01-01  2023-12-31   
17  2024-04-08 22:41:48.000000                  total  2023-10-01  2023-10-31   
18  2024-04-08 22:41:48.000000                  total  2023-11-01  2023-11-30   
19  2024-04-08 22:41:48.000000                  total  2023-12-01  2023-12-31   
20  2024-04-08 22:41:48.000000  ferc:GasUtilityMember  2022-01-01  2022-12-31   
21  2025-04-14 20:52:59.000000                  total  2024-01-01  2024-12-31   
22  2025-04-14 20:52:59.000000                  total  2023-01-01  2023-12-31   
23  2025-04-14 20:52:59.000000                  total  2024-10-01  2024-10-31   
24  2025-04-14 20:52:59.000000    

In [27]:
df_wbi_energy_transmission

entity_id                                        filing_name  \
0    C000111  WBI_Energy_Transmission,_Inc._form2_Q4_1650061854   
1    C000111  WBI_Energy_Transmission,_Inc._form2_Q4_1650061854   
2    C000111  WBI_Energy_Transmission,_Inc._form2_Q4_1650061854   
3    C000111  WBI_Energy_Transmission,_Inc._form2_Q4_1650061854   
4    C000111  WBI_Energy_Transmission,_Inc._form2_Q4_1650061854   
5    C000111  WBI_Energy_Transmission,_Inc._form2_Q4_1650061854   
6    C000111  WBI_Energy_Transmission,_Inc._form2_Q4_1650061854   
7    C000111  WBI_Energy_Transmission,_Inc._form2_Q4_1681500425   
8    C000111  WBI_Energy_Transmission,_Inc._form2_Q4_1681500425   
9    C000111  WBI_Energy_Transmission,_Inc._form2_Q4_1681500425   
10   C000111  WBI_Energy_Transmission,_Inc._form2_Q4_1681500425   
11   C000111  WBI_Energy_Transmission,_Inc._form2_Q4_1681500425   
12   C000111  WBI_Energy_Transmission,_Inc._form2_Q4_1681500425   
13   C000111  WBI_Energy_Transmission,_Inc._form2_Q4_1681500425   
14   C000111  WBI_Energy_Transmission,_Inc._form2_Q4_1713220065   
15   C000111  WBI_Energy_Transmission,_Inc._form2_Q4_1713220065   
16   C000111  WBI_Energy_Transmission,_Inc._form2_Q4_1713220065   
17   C000111  WBI_Energy_Transmission,_Inc._form2_Q4_1713220065   
18   C000111  WBI_Energy_Transmission,_Inc._form2_Q4_1713220065   
19   C000111  WBI_Energy_Transmission,_Inc._form2_Q4_1713220065   
20   C000111  WBI_Energy_Transmission,_Inc._form2_Q4_1713220065   
21   C000111  WBI_Energy_Transmission,_Inc._form2_Q4_1744757061   
22   C000111  WBI_Energy_Transmission,_Inc._form2_Q4_1744757061   
23   C000111  WBI_Energy_Transmission,_Inc._form2_Q4_1744757061   
24   C000111  WBI_Energy_Transmission,_Inc._form2_Q4_1744757061   
25   C000111  WBI_Energy_Transmission,_Inc._form2_Q4_1744757061   
26   C000111  WBI_Energy_Transmission,_Inc._form2_Q4_1744757061   
27   C000111  WBI_Energy_Transmission,_Inc._form2_Q4_1744757061   

              publication_time      utility_type_axis  start_date    end_date  \
0   2022-04-15 22:30:54.000000                  total  2021-12-01  2021-12-31   
1   2022-04-15 22:30:54.000000                  total  2020-01-01  2020-12-31   
2   2022-04-15 22:30:54.000000                  total  2021-01-01  2021-12-31   
3   2022-04-15 22:30:54.000000                  total  2021-11-01  2021-11-30   
4   2022-04-15 22:30:54.000000  ferc:GasUtilityMember  2021-01-01  2021-12-31   
5   2022-04-15 22:30:54.000000  ferc:GasUtilityMember  2020-01-01  2020-12-31   
6   2022-04-15 22:30:54.000000                  total  2021-10-01  2021-10-31   
7   2023-04-14 19:27:05.000000                  total  2021-01-01  2021-12-31   
8   2023-04-14 19:27:05.000000  ferc:GasUtilityMember  2022-01-01  2022-12-31   
9   2023-04-14 19:27:05.000000  ferc:GasUtilityMember  2021-01-01  2021-12-31   
10  2023-04-14 19:27:05.000000                  total  2022-12-01  2022-12-31   
11  2023-04-14 19:27:05.000000                  total  2022-10-01  2022-10-31   
12  2023-04-14 19:27:05.000000                  total  2022-11-01  2022-11-30   
13  2023-04-14 19:27:05.000000                  total  2022-01-01  2022-12-31   
14  2024-04-15 22:27:45.000000                  total  2023-01-01  2023-12-31   
15  2024-04-15 22:27:45.000000                  total  2022-01-01  2022-12-31   
16  2024-04-15 22:27:45.000000  ferc:GasUtilityMember  2023-01-01  2023-12-31   
17  2024-04-15 22:27:45.000000  ferc:GasUtilityMember  2022-01-01  2022-12-31   
18  2024-04-15 22:27:45.000000                  total  2023-10-01  2023-10-31   
19  2024-04-15 22:27:45.000000                  total  2023-11-01  2023-11-30   
20  2024-04-15 22:27:45.000000                  total  2023-12-01  2023-12-31   
21  2025-04-15 22:44:21.000000                  total  2024-01-01  2024-12-31   
22  2025-04-15 22:44:21.000000                  total  2023-01-01  2023-12-31   
23  2025-04-15 22:44:21.000000  ferc:GasUtilityMember  2024-01-01  2024-12-31   
24  2025-04-15 22:44:21.000000  fe

## Conclusion

We were able to successfully extract a complete FERC Form 2 Statement of Income from the PUDL SQL database and were able to build a custom extract function. This function can now be used to extract Form 2 Income Statement data as needed.